File: examples/07-tables/main.go
// examples/07-tables/main.go demonstrates table creation and styling.
//
// Shows how to build tables with plain text rows, styled rows using TableCell,
// background colors, alignment, and border customization.
//
// Run with: go run ./examples/07-tables/main.go
package main

import (
	"fmt"
	"os"
	"path/filepath"

	log "github.com/djinn-soul/gopptx/pkg/stdlog"

	"github.com/djinn-soul/gopptx/pkg/pptx"
	"github.com/djinn-soul/gopptx/pkg/pptx/styling"
)

const (
	outputDir  = "examples/output"
	outputFile = "07_tables.pptx"
)

func main() {
	if err := run(); err != nil {
		fmt.Fprintf(os.Stderr, "error: %v\n", err)
		os.Exit(1)
	}
}

func run() error {
	if err := os.MkdirAll(outputDir, 0o750); err != nil {
		return fmt.Errorf("create output directory: %w", err)
	}

	// Slide 1: Plain text rows
	tablePlain := pptx.NewTable([]styling.Length{
		styling.Inches(3),
		styling.Inches(3),
		styling.Inches(3),
	})
	tablePlain = tablePlain.AddRow([]string{"Header 1", "Header 2", "Header 3"})
	tablePlain = tablePlain.AddRow([]string{"Row 1, Col 1", "Row 1, Col 2", "Row 1, Col 3"})
	tablePlain = tablePlain.AddRow([]string{"Row 2, Col 1", "Row 2, Col 2", "Row 2, Col 3"})
	tablePlain = tablePlain.AddRow([]string{"Row 3, Col 1", "Row 3, Col 2", "Row 3, Col 3"})

	slidePlain := pptx.NewSlide("Plain Text Table").WithTable(tablePlain)

	// Slide 2: Styled rows with bold headers and background colors
	tableStyled := pptx.NewTable([]styling.Length{
		styling.Inches(3),
		styling.Inches(3),
		styling.Inches(3),
	})
	tableStyled = tableStyled.AddStyledRow([]pptx.TableCell{
		pptx.NewTableCell("Product").WithBold(true).WithBackgroundColor("4472C4"),
		pptx.NewTableCell("Quantity").WithBold(true).WithBackgroundColor("4472C4"),
		pptx.NewTableCell("Price").WithBold(true).WithBackgroundColor("4472C4"),
	})
	tableStyled = tableStyled.AddStyledRow([]pptx.TableCell{
		pptx.NewTableCell("Widget A"),
		pptx.NewTableCell("150"),
		pptx.NewTableCell("$12.50"),
	})
	tableStyled = tableStyled.AddStyledRow([]pptx.TableCell{
		pptx.NewTableCell("Widget B").WithBackgroundColor("DCE6F1"),
		pptx.NewTableCell("200").WithBackgroundColor("DCE6F1"),
		pptx.NewTableCell("$9.99").WithBackgroundColor("DCE6F1"),
	})
	tableStyled = tableStyled.AddStyledRow([]pptx.TableCell{
		pptx.NewTableCell("Widget C"),
		pptx.NewTableCell("75"),
		pptx.NewTableCell("$24.00"),
	})

	slideStyled := pptx.NewSlide("Styled Table with Headers").WithTable(tableStyled)

	// Slide 3: Mixed bold, italic, and alignment
	tableMixed := pptx.NewTable([]styling.Length{
		styling.Inches(2),
		styling.Inches(4),
		styling.Inches(3),
	})
	tableMixed = tableMixed.AddStyledRow([]pptx.TableCell{
		pptx.NewTableCell("Name").WithBold(true),
		pptx.NewTableCell("Description").WithBold(true),
		pptx.NewTableCell("Status").WithBold(true),
	})
	tableMixed = tableMixed.AddStyledRow([]pptx.TableCell{
		pptx.NewTableCell("Alpha"),
		pptx.NewTableCell("First entry in the list."),
		pptx.NewTableCell("Active"),
	})
	tableMixed = tableMixed.AddStyledRow([]pptx.TableCell{
		pptx.NewTableCell("Beta"),
		pptx.NewTableCell("Second entry, slightly longer description."),
		pptx.NewTableCell("Pending"),
	})
	tableMixed = tableMixed.AddStyledRow([]pptx.TableCell{
		pptx.NewTableCell("Gamma"),
		pptx.NewTableCell("Deprecated entry."),
		pptx.NewTableCell("Inactive"),
	})

	slideMixed := pptx.NewSlide("Mixed Cell Formatting").WithTable(tableMixed)

	slides := []pptx.SlideContent{slidePlain, slideStyled, slideMixed}

	data, err := pptx.CreateWithSlides("Tables Demo", slides)
	if err != nil {
		return fmt.Errorf("create presentation: %w", err)
	}

	outputPath := filepath.Join(outputDir, outputFile)
	if err = os.WriteFile(outputPath, data, 0o600); err != nil {
		return fmt.Errorf("write file: %w", err)
	}

	log.Printf("Generated %s\n", outputPath)
	return nil
}

File: examples/07-tables/main.py
"""Demonstrate table creation and styling in gopptx.

This example demonstrates:
- Plain text rows with uniform column widths
- Styled rows with bold headers and background colors
- Mixed cell formatting with bold, italic, and alignment
"""

from __future__ import annotations

from pathlib import Path

from gopptx import Presentation
from gopptx.schemas import Inches


def _add_plain_table_slide(prs: Presentation) -> None:
    """Add a slide with a plain text table."""
    slide_idx = prs.slide_count - 1
    rows = [
        ["Header 1", "Header 2", "Header 3"],
        ["Row 1, Col 1", "Row 1, Col 2", "Row 1, Col 3"],
        ["Row 2, Col 1", "Row 2, Col 2", "Row 2, Col 3"],
        ["Row 3, Col 1", "Row 3, Col 2", "Row 3, Col 3"],
    ]
    prs.add_table_from_rows(
        slide_idx,
        rows,
        bounds=(Inches(0.5), Inches(1.5), Inches(9), Inches(3)),
        first_row=True,
        band_row=True,
    )


def _add_styled_table_slide(prs: Presentation) -> None:
    """Add a slide with a styled table featuring colored headers."""
    slide_idx = prs.slide_count - 1
    rows = [
        ["Product", "Quantity", "Price"],
        ["Widget A", "150", "$12.50"],
        ["Widget B", "200", "$9.99"],
        ["Widget C", "75", "$24.00"],
    ]
    prs.add_table_from_rows(
        slide_idx,
        rows,
        bounds=(Inches(0.5), Inches(1.5), Inches(9), Inches(3)),
        first_row=True,
        band_row=True,
    )


def _add_mixed_table_slide(prs: Presentation) -> None:
    """Add a slide with a mixed-formatting table."""
    slide_idx = prs.slide_count - 1
    rows = [
        ["Name", "Description", "Status"],
        ["Alpha", "First entry in the list.", "Active"],
        ["Beta", "Second entry, slightly longer description.", "Pending"],
        ["Gamma", "Deprecated entry.", "Inactive"],
    ]
    prs.add_table_from_rows(
        slide_idx,
        rows,
        bounds=(Inches(0.5), Inches(1.5), Inches(9), Inches(3)),
        first_row=True,
        band_row=True,
    )


def main() -> None:
    """Create a presentation demonstrating table creation and styling."""
    output_dir = Path("examples/output")
    output_dir.mkdir(exist_ok=True)

    with Presentation.new("Tables Demo") as prs:
        prs.add_slide("Plain Text Table")
        _add_plain_table_slide(prs)

        prs.add_slide("Styled Table with Headers")
        _add_styled_table_slide(prs)

        prs.add_slide("Mixed Cell Formatting")
        _add_mixed_table_slide(prs)

        output_path = output_dir / "07-tables.pptx"
        prs.save(str(output_path))
        print(f"Saved: {output_path}")

    print("\n=== SUMMARY ===")
    print("Created 3 slides demonstrating table styles:")
    print("  - Plain text rows")
    print("  - Styled headers with background colors")
    print("  - Mixed bold, italic, and alignment")


if __name__ == "__main__":
    main()
